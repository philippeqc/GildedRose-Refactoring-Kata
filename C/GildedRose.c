#include "GildedRose.h"
#include <stdio.h>
#include <string.h>

Item *init_item(Item *item, const char *name, int sellIn, int quality) {
  item->sellIn = sellIn;
  item->quality = quality;
  item->name = strdup(name);

  return item;
}

extern char *print_item(char *buffer, Item *item) {
  sprintf(buffer, "%s, %d, %d", item->name, item->sellIn, item->quality);
}

typedef enum {
  ITEM_COMMON,
  ITEM_AGED_BRIE,

  ITEM_SULFURAS,
  ITEM_BACKSTAGE_PASSES,
  ITEM_CONJURED
} item_type;

item_type getItemType(Item *pItem) {
  if (strcmp(pItem->name, "Aged Brie") == 0) {
    return ITEM_AGED_BRIE;
  }
  if (strcmp(pItem->name, "Backstage passes to a TAFKAL80ETC concert") == 0) {
    return ITEM_BACKSTAGE_PASSES;
  }
  if (strcmp(pItem->name, "Sulfuras, Hand of Ragnaros") == 0) {
    return ITEM_SULFURAS;
  }
  if (strcmp(pItem->name, "Conjured") == 0) {
    return ITEM_CONJURED;
  }

  return ITEM_COMMON;
}

void update_item_quality(Item *pItem) {
  switch (getItemType(pItem)) {
  case ITEM_COMMON:
  case ITEM_CONJURED: {
    if (pItem->quality > 0) {
      if (pItem->sellIn == 0) {
        pItem->quality = pItem->quality - 2;
      } else {
        pItem->quality = pItem->quality - 1;
      }
    }
    break;
  }
  case ITEM_AGED_BRIE: {
    if (pItem->sellIn < 0) {
      pItem->quality = pItem->quality + 2;
    } else {
      pItem->quality = pItem->quality + 1;
    }
    break;
  }
  case ITEM_BACKSTAGE_PASSES: {
    if (pItem->quality < 50) {
      if (pItem->sellIn == 0) {
        pItem->quality = 0;
      } else if (pItem->sellIn < 6) {
        pItem->quality = pItem->quality + 3;
      } else if (pItem->sellIn < 11) {
        pItem->quality = pItem->quality + 2;
      } else {
        pItem->quality = pItem->quality + 1;
      }
    }
    break;
  }
  case ITEM_SULFURAS: {
    return;
    break;
  }
  }

  pItem->sellIn = pItem->sellIn - 1;

  if (pItem->quality > 50) {
    pItem->quality = 50;
  }
  if (pItem->quality < 0) {
    pItem->quality = 0;
  }
}

void update_quality(Item items[], int size) {
  int i;

  for (i = 0; i < size; i++) {
    update_item_quality(&items[i]);
  }
}
