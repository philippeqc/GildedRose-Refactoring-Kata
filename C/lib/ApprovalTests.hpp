#include "ApprovalTests.v.10.12.2.hpp"
