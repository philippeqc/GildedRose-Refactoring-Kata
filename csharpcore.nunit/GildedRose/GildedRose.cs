using System;
using System.Collections.Generic;

namespace GildedRoseKata;

public class Sulfuras
{
    private Item m_item;
    public Sulfuras(Item item) { m_item = item; }
    public void UpdateQuality() { }
}

public class AgedBrie
{
    private Item m_item;
    public AgedBrie(Item item) { m_item = item; }
    public void UpdateQuality()
    {
        if (m_item.Quality < 50)
        {
            m_item.Quality += 1;
        }
    }
}

public class GildedRose
{
    private readonly IList<Item> _items;

    public GildedRose(IList<Item> items)
    {
        _items = items;
    }

    public void UpdateQuality()
    {
        foreach (var item in _items)
        {
            UpdateItemQuality(item);
        }
    }

    public void UpdateItemQuality(Item item)
    {
        if (item.Name == "Sulfuras, Hand of Ragnaros")
        {
            Sulfuras sulfuras = new Sulfuras(item);
            sulfuras.UpdateQuality();
            return;
        }

        item.SellIn -= 1;

        if (item.Name == "Aged Brie")
        {
            AgedBrie agedBrie = new AgedBrie(item);
            agedBrie.UpdateQuality();
            return;
        }

        if (item.Name == "Backstage passes to a TAFKAL80ETC concert")
        {
            if (item.SellIn <= 0)
            {
                item.Quality = 0;
            }
            else if (item.SellIn <= 5)
            {
                item.Quality += 3;
            }
            else if (item.SellIn <= 10)
            {
                item.Quality += 2;
            }
            else
            {
                item.Quality += 1;
            }
            return;
        }

        if(item.Name == "Conjured") {
            if (item.SellIn <= 0)
            {
                item.Quality = Math.Max(item.Quality - 4, 0);
            }
            else
            {
                item.Quality = Math.Max(item.Quality - 2, 0);
            }
            return;
        }

        if (item.SellIn <= 0)
        {
            item.Quality = Math.Max(item.Quality - 2, 0);
        }
        else
        {
            item.Quality = Math.Max(item.Quality - 1, 0);
        }
        return;
    }
}