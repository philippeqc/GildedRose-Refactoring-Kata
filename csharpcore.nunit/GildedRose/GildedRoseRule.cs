using System.Collections.Generic;
using GildedRoseKata.Domain.Rule;
using GildedRoseKata.Rule;

namespace GildedRoseKata;

public class GildedRoseRule
{
    public IDictionary<string, ItemRule> rules { get; }
    private ItemRule defaultRule;
    public ItemRule Default { get { return defaultRule; } }

    public GildedRoseRule()
    {
        rules = new Dictionary<string, ItemRule>
        {
            { "Backstage passes to a TAFKAL80ETC concert", BackstagePassesRule.Rule() },
            { "Sulfuras, Hand of Ragnaros", SulfurasRule.Rule() },
            { "Aged Brie", AgedBrieRule.Rule() }
        };
        defaultRule = DefaultRule.Rule();
    }
}