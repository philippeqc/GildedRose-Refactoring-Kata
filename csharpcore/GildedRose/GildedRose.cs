using System.Collections.Generic;
using GildedRoseKata.Rule;

namespace GildedRoseKata;

public class GildedRose
{
    private readonly IList<Item> _items;
    private readonly IDictionary<string, ItemRule> _rules;
    private readonly ItemRule _defaultRule;

    public GildedRose(IList<Item> items, IDictionary<string, ItemRule> rules, ItemRule defaultRule)
    {
        _items = items;
        _rules = rules;
        _defaultRule = defaultRule;
    }

    private void UpdateItem(Item item)
    {
        if (!_rules.TryGetValue(item.Name, out ItemRule itemRule))
        {
            itemRule = _defaultRule;
        }
        itemRule.DoRule(item);
    }

    public void UpdateQuality()
    {
        foreach (var item in _items)
        {
            UpdateItem(item);
        }
    }
}
