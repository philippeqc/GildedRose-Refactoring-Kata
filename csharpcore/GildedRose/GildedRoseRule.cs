using System.Collections.Generic;
using GildedRoseKata.Domain.Rule;
using GildedRoseKata.Rule;

namespace GildedRoseKata;

public class GildedRoseRule
{
    public IDictionary<string, ItemRule> rules { get; }
    private ItemRule defaultRule;
    public ItemRule DefaultRule { get { return defaultRule; } }

    public GildedRoseRule()
    {
        rules = new Dictionary<string, ItemRule>
        {
            { "Backstage passes to a TAFKAL80ETC concert", new BackstagePassesRule().rule },
            { "Sulfuras, Hand of Ragnaros", new QualityConstantRule(80) },
            { "Aged Brie", new AgedBrieRule().rule }
        };
        defaultRule = new DefaultRule().rule;
    }
}