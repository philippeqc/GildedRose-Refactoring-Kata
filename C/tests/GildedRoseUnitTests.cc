#include "ApprovalTests.hpp"
#include <catch2/catch_test_macros.hpp>

using namespace ApprovalTests;

extern "C" {
#include "../GildedRose.h"
}

TEST_CASE("AFirstTest") {
  Item items[1];
  init_item(items, "Foo", 0, 0);
  update_quality(items, 1);

  Approvals::verify(items[0], [](const Item &b, std::ostream &os) {
    os << "name: " << b.name << ", sellIn: " << b.sellIn
       << ", quantity: " << b.quality;
  });
}

void ostreamItem(const Item &b, std::ostream &os) {
  os << "name: " << b.name << ", sellIn: " << b.sellIn
     << ", quantity: " << b.quality;
};

TEST_CASE("TestGildedRoseGroup", "FirstTest") {
  Item items[6];
  int last = 0;

  init_item(items + last++, "+5 Dexterity Vest", 10, 20);
  init_item(items + last++, "Aged Brie", 2, 0);
  init_item(items + last++, "Elixir of the Mongoose", 5, 7);
  init_item(items + last++, "Sulfuras, Hand of Ragnaros", 0, 80);
  init_item(items + last++, "Backstage passes to a TAFKAL80ETC concert", 15,
            20);
  init_item(items + last++, "Conjured Mana Cake", 3, 6);

  SECTION(
      "At the end of each day our system lowers both values for every item") {
    // arrange
    Item vest = items[0];

    // act
    update_quality(items, last);

    // assert
    Item actual = items[0];
    Approvals::verify(actual, ostreamItem);
  }

  SECTION("The product looses 1 quality per day until it reaches the sell by "
          "date") {
    // arrange
    Item vest = items[0];

    // act
    for (int i = vest.sellIn; i > 0; i--) {
      update_quality(items, last);
    }

    // assert
    // REQUIRE(vest.quality - vest.sellIn == items[0].quality);
    Item actual = items[0];
    Approvals::verify(actual, ostreamItem);

    SECTION(
        "Once the sell by date has passed, Quality degrades twice as fast") {
      // Item vest = items[0];

      // act
      update_quality(items, last);

      // assert
      Item actual = items[0];
      Approvals::verify(actual, ostreamItem);
      // Or could be
      // REQUIRE(vest.quality - 2 == actual.quality);

      SECTION("The quality can degrade to 0") {
        // arrange
        Item vest = items[0];

        // act
        int daysLeft = (vest.quality + 1) / 2;
        for (int i = daysLeft; i > 0; i--) {
          update_quality(items, last);
        }

        // assert
        REQUIRE(0 == items[0].quality);

        SECTION("The Quality of an item is never negative") {

          // act
          update_quality(items, last);

          // assert
          Item actual = items[0];
          REQUIRE(0 == actual.quality);
          // This fails due to a too long file name
          // Approvals::verify(actual, ostreamItem);
        }
      }
    }
  }

  SECTION("Aged Brie actually increases in Quality the older it gets") {
    Item brie = items[1];
    REQUIRE(strcmp("Aged Brie", brie.name) == 0);

    
  }
}