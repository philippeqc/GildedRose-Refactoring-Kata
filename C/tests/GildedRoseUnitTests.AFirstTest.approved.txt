name: Foo, sellIn: -1, quantity: 0
